### ISSUE TITLE
Cube and ObservationSet best practices?

### ISSUE THREAD

**Issue by lapidus** [zazuko/rdf-cube-schema#23]
My understanding thus far is that a cube:Cube can hold any number of dimensions and measures as long as the observations conform to the same dimensional structure.

Meaning, a Cube could look like this? (flattened)

```
country, gender, age_group, population (measure), income_per_person (measure)
```

And an ObservationSet could for example be restricted to one measure like this:

```
country, gender, age_group, population
sweden, male, 10_20y, 9999
```

An Observation could either store multiple measures ...

```
country, gender, age_group, population, income_per_person
sweden, male, 10_20y, 9999, 9999
```

Or a single one ... 

```
country, gender, age_group, population
sweden, male, 10_20y, 9999
```

Now, to my questions:

**a) Store lots of measures on one cube?**
If we have 100 measures on country--year combination, is there any reason not to put them in the same cube? It seems to help from a categorization and join perspective ... 

**b) Store lots of measures on each observation vs splitting it up?** 
Given 100 measures on country--year, would it make sense to store all measures on each observation?

{uuid} a :Observation
  :country "usa"; 
  :year 2021;
  :population 9999;
  :income_per_person 9999;
  + 98 ...
 
Or is this when one runs into problems with sparse data and expensive { OPTIONALs }?  

On the other hand, isn't it unnecessarily redundant to create separate observations linked to separate ObservationSets?

:populationByCountryYear a cube:ObservationSet;
   :observation :obs1

:obs1 a :Observation
  :country "usa"; 
  :year 2021;
  :population 9999 .

:obs99 a :Observation
  :country "usa"; 
  :year 2021;
  :income_per_person 9999;

Basically I wonder about the value of the ObservationSet overhead as well as the risk of redundant Observations when more data can be captured in each. 

**c) Integrity of Observations – how to upsert data in a safe manner?**
Whether single Observations feature all measures or are limited to a few of the cube's measures, how does one maintain integrity when updating observations? For example in the example above it would make sense to check if there's an Observation with country=usa AND year=2021 and upsert values to that when adding new measures.

Thank you :)

**Comment by ktk** [comment]
1. Yes, you can have any number of dimensions and measures as long as it stays the same in that particular cube.
2. Yes, the `ObservationSet` is basically an index, and if it is useful, a shortcut into a subset of the obervations. This also means you would absolutely add a new year on the same cube as the year before. If you want an additional index for that particular year, you simply create an additional `ObservationSet`.
3. An `Observation` can have as many measures as you like, it depends on what you want to do with the data. We usually just do that for things that are measured together, so I wouldn't do multiple measures in one cube if they have nothing to do with each other, like they are measured independently of each other.

So:
a. As mentioned in 3. We usually split the cubes if it's not measured at the same time. I get the point that the rest of the dimensions might be the same but we would most likely do different cubes anyway. I think I would only do it if I query/show all of them at the same time/in the same diagram but I don't think that would make any sense from a visualization POV.
b. is the same question or am I missing something? Obviously the measures are assigned to an `Observation` but they would belong to the same cube and the same cube structure.
c. Not sure if I get that, you would simply add a new `Observation` with whatever things you have defined in that cube as dimensions/measures. The integrity could be checked by the SHACL shape attached to it.

**Comment by lapidus** [comment]
Thank you for clarifying.

> but I don't think that would make any sense from a visualization POV.

We have some pages that would benefit from displaying let's say 20 measures for the same year/country combination so it might make sense to colocate them in the same observation to get them in one simple call. And overall, as we have full control over our cubes (and prioritize logical partitioning over flexibility), it seems like an unnecessary abstraction for us to think in terms of ObservationSets as we simply need to know what dimensional combinations we intend to store and what measures we have for those keys. Does that make sense? I can understand that Cubes+ObservationSets make a lot of sense if you have multiple data providers creating cubes, but if we intend to harmonize the data and just need a logical storage, it seems we get away with Cubes and Observations only.
 
> Not sure if I get that, you would simply add a new Observation with whatever things you have defined in that cube as dimensions/measures

The case I am thinking about follows on the above. Let's say we have this in one cube:

observationID, country, gender, year, population
obs1, sweden, male, 2020, 999

We now have a script that is responsible for adding one new measure to this cube. 

observationID, country, gender, year, population, **income**
obs1, sweden, male, 2020, 999, **888**

And one script that is responsible for adding observations for one measure such as population.

observationID, country, gender, year, **population**, income
obs1, sweden, male, 2020, 999, 888
obs2, sweden, male, 2021, **996**, null

To allow these more granular operations, that target particular observations of the cube, I was wondering if you have any particular strategies? For example if each observation's ID was a hash of the dimensions, it seems one could relatively easily "diff" existing and new data and create or append to observations?

**Comment by ktk** [comment]
> Does that make sense? I can understand that Cubes+ObservationSets make a lot of sense if you have multiple data providers creating cubes, but if we intend to harmonize the data and just need a logical storage, it seems we get away with Cubes and Observations only.

ok I see that point. To be clear, the cube model does not stop you doing that at all so if you are comfortable with the data that way, go for it. And yes it is cheaper to query it most likely.

> The case I am thinking about follows on the above. Let's say we have this in one cube:
> [...]
> To allow these more granular operations, that target particular observations of the cube, I was wondering if you have any particular strategies? For example if each observation's ID was a hash of the dimensions, it seems one could relatively easily "diff" existing and new data and create or append to observations?

The approach depends on your pipeline I would say. If you do it outside a triplestore/SPARQL endpoint, your pipeline logic needs to take care of it and needs to make sure that you add empty rows to every existing observation that is there. A hash of them somewhere might be a way to do that, yes.

It is very easy to add data in RDF as the store will take care of the "schema" to speak in RDB terms. With standards like SHACL you could also validate that you took care of everything and didn't miss anything.

If it's already in a triplestore you might do that insert with SPARQL UPDATE protocol as well so it should be relatively easy to add that to all observations where it is missing.

**Comment by l00mi** [comment]
No further questions, closing for now @lapidus please open up again if still uncertainties are around.
